torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -333,13 +333,15 @@ ProcessGroupGloo::SendWork::SendWork(
 
 bool ProcessGroupGloo::SendWork::wait() {
   bool sendCompleted = false;
-  std::unique_lock<std::mutex> lock(mutex_);
   try {
     sendCompleted = buffer_->waitSend();
   } catch (...) {
+    // Lock to write exception_
+    std::unique_lock<std::mutex> lock(mutex_);
     exception_ = std::current_exception();
```

Comment:
Can we avoid having two locks here by having a function local exception variable and only acquire the lock after the catch? Will it be better if we have a finally block and acquire lock there? sth like:

```c++
try {
  sendCompleted = buffer_->waitSend();
} catch (...) {
  exception = std::current_exception();
} finally {
  std::lock_guard<std::mutex> lock(mutex_);
  completed_ = true;
  if (exception) {
    std::swap(exception, exception_);
    std::rethrow_exception(exception_);
  }
}

return sendCompleted;
```

Reply:
I think we can add the function local variable, although C++ does not support `finally` for the 2nd suggestion.